strongbox-web-core/src/main/java/org/carlspring/strongbox/security/vote/CustomAccessDecisionVoter.java

```diff
@@ -0,0 +1,123 @@
+package org.carlspring.strongbox.security.vote;
+
+import org.carlspring.strongbox.security.user.SpringSecurityUser;
+import org.carlspring.strongbox.users.domain.Features;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.aop.ProxyMethodInvocation;
+import org.springframework.security.access.AccessDecisionVoter;
+import org.springframework.security.access.ConfigAttribute;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.web.FilterInvocation;
+
+/**
+ * Customization of {@link AccessDecisionVoter} according to custom access strategies for users
+ * like per-repository access etc.
+ *
+ * @author Alex Oreshkevich
+ * @see https://youtrack.carlspring.org/issue/SB-603
+ */
+public class CustomAccessDecisionVoter
+        implements AccessDecisionVoter, AuthenticationProvider
+{
+
+    private static final Logger logger = LoggerFactory.getLogger(CustomAccessDecisionVoter.class);
+
+    private Authentication authentication;
+
+    @Override
+    public boolean supports(ConfigAttribute attribute)
+    {
+        // don't need any specific configuration attributes
+        return true;
+    }
+
+    @Override
+    public int vote(Authentication authentication,
+                    Object object,
+                    Collection collection)
+    {
+        int vote = ACCESS_ABSTAIN;
+        this.authentication = authentication;
+
+        SpringSecurityUser user = (SpringSecurityUser) authentication.getPrincipal();
+        Features features = user.getFeatures();
+        if (features == null)
+        {
+            return vote;
+        }
+
+        // save request url during first authentication phase
+        // parse repositoryId from pre-saved during second authentication phase
+        if (object instanceof FilterInvocation)
+        {
+            user.setUrl(((FilterInvocation) object).getRequestUrl());
```

Comment:
Kind of opening door in opposite direction but I agree here with you. It will look more readable.